### ISSUE TITLE
Docker Support

### ISSUE THREAD

**Issue by sschlein** [beyondcode/tinkerwell-community#2]
Connect to docker containers

**Comment by gilbitron** [comment]
In particular, it would be great to support https://github.com/shipping-docker/vessel

**Comment by Phunky** [comment]
If we can manually enter the command for running php that should enable docker support really crudely and quickly.

**Comment by mathieutu** [comment]
I've created a more generic issue about a way to connect to "custom connections" (docker, VM, cloud, etc.): https://github.com/beyondcode/tinkerwell-suggestions/issues/46

**Comment by tjespers** [comment]
Yesterday i toyed around with the possibility of having the PHP connection settings be customisable & extendable in a similar way as we can already implement custom tinkerwell drivers. As a result of that i decided to share my take on things here to see if this is something that might be deemed useful to others as well.

What i was looking for is a custom connection on a per project basis as all my projects use different docker / docker-compose stacks & containers. As an experiment i've created a simple bash proxy which routes all tinkerwell calls to a custom PHP connection based on the presence of a .tinkerwell/cli.interpreter file in my project,

Code & examples can be found at tjespers/tinkerwell-interpreters

**Comment by richbreton** [comment]
seems like this could be implemented and treated much like ssh (not using ssh) just have tinkerwell have a field like the ssh settings page where you fill out a container name ant then tinkerwell could shell into docker and then run everything from that shell like it does with ssh: 

docker exec -it $container_name bash

or it could prepend every command it runs with docker exec -it $container_name php artisan blah

I have seen some suggestions about just installing sshd on your container, this is a bad idea and is not "the docker way" we push our containers to prod and none of them have ssh enabled. 

I own a license and we have been tryiing to test this for our department to buy licenses . thanks

**Comment by gpressutto5** [comment]
I believe the solution suggested by @richbreton would work well and it would only require the container name

**Comment by michabbb** [comment]
I would buy a license, but without docker support, I cannot use it 😥

**Comment by michael-mcmullen** [comment]
I own a license but having to install ssh on my docker images and keys etc just for development is an extra step that makes it very unpleasant currently.

**Comment by tyteck** [comment]
The same as Michael. In fact I was about to buy it, then I made a docker search.

The way I'm using docker is running a `docker-compose up -d` with my laravel app within.
Once running I'm using artisan commands like this
`docker exec -it <laravelContainerName> php artisan`.
for route list, it is doing
`docker exec -it <laravelContainerName> php artisan route:list`
If tinker is able to run php with a command line prefix. It should be good.
Cheers

**Comment by jehadassaf** [comment]
Same here. Once there's Docker support in Tinkerwell I will hit that Buy button so hard my mouse will likely explode.

**Comment by negoziator** [comment]
I already bought it!  My mouse didn't explode :eyes: 
- but when docker is supported it will use it so hard, my mouse will likely explode too! :boom: :whale:

**Comment by Obvinyg** [comment]
I found a way to make Tinkerwell work with Docker. You have to create your own PHP executable.

An example: *php.sh*
```
#!/bin/bash
cat /tmp/data.txt | docker-compose -f /home/daniel/development/project-xyz/docker-compose.yml exec -T develop php artisan tinker
```

Now to make it work you have to set your path for docker-compose manually (because the command is executed from a different directory).

You also need to replace develop with whatever the name of your docker container is.

Go to preferences -> advanced and set the PHP binary to point to php.sh (after making it executable with `chmod +x php.sh`) and it should work :)

**Comment by adrian-schnell** [comment]
> I found a way to make Tinkerwell work with Docker (tested with Tinkerwell 2.7.0).
> 
> You have to create your own PHP executable.
> 
> ### php.sh
> ```
> #!/bin/bash
> cat $2 | docker-compose -f $3/docker-compose.yml exec -T dev php artisan tinker
> ```
> 
> 1. Copy my command to a php.sh file.
> 2. Make sure that the working directory is pointed to your app (and update the path to docker-compose.yml if neccesary)
> 3. Update the name of the container (**dev** in my example)
> 4. Make the script executable (`chmod +x php.sh`)
> 5. Open Tinkerwell and navigate to Preferences -> Advanced
> 6. Set the newly created php.sh file to be your PHP binary.
> 7. ????
> 8. Profit
> 
> _The detected PHP version in the Preferences->Advanced window will not display correctly after you do this.. Oh well_

I'm getting the error `line 2: docker-compose: command not found` - my docker-compose.yml and the php.sh are both in the root of the laravel application. I just changed "dev" to "app" to point to my own container.

**Comment by gpressutto5** [comment]
I really don't mind paying for good software and tinkerwell definitely is worth it, but, man, if it were open source we would have docker support months ago...

**Comment by spielfigur** [comment]
> I'm getting the error `line 2: docker-compose: command not found` - my docker-compose.yml and the php.sh are both in the root of the laravel application. I just changed "dev" to "app" to point to my own container.
> 
> Inside the local shell, `docker-compose -f ./docker-compose.yml exec app php artisan tinker` is working fine.

+1
got the same problem 😕

**Comment by TTiger** [comment]
Replace docker-compose with the full path to your local executable.   For example, mine is:
/usr/local/bin/docker-compose

**Comment by semiherdogan** [comment]
For single line code its ok. But for multiline it throws syntax error.

Any idea how to solve this?

This one is getting results from database.
```
\App\Models\User::get();
```

This one throws exception.
```
\App\Models\User::query()
  ->get();
```

**Comment by negoziator** [comment]
> For single line code its ok. But for multiline it throws syntax error.
> 
> Any idea how to solve this?
> 
> This one is getting results from database.
> 
> ```
> \App\Models\User::get();
> ```
> 
> This one throws exception.
> 
> ```
> \App\Models\User::query()
>   ->get();
> ```

That's really weird you experience that - I just tested the exact same query as you sent. It seems to work just fine. *

Maybe it's your version of tinker in your project that needs to be updated?

**Comment by semiherdogan** [comment]
My project's laravel version is `6.18.3` and from `composer.lock` laravel/tinker  is `11 months ago | ad571aa` .
Should i update laravel/tinker? and i'm not sure is it safe to update it without laravel update?

my php.sh file is

```
#!/bin/bash
PATH=$PATH:/usr/local/bin
cat $2 | docker-compose -f $3/docker-compose.yml exec -T app php artisan tinker
```

**Comment by millsoft** [comment]
I just bought tinkerwell and wanted to know how to use it for project running in docker - as most of my projects run in docker. The provided solution here (by creating own php.sh) works for me fine but it would be nice to have a built in docker feature in tinkerwell.

**Comment by mikeerickson** [comment]
@millsoft sounds like this is finally on the sooner rather than later feature according to @mpociot  hoping that further releases focus on addressing issues as opposed to adding more features.

**Comment by gpressutto5** [comment]
I think they should say on the website that it doesn't support docker. I know they never said it does, but I'm sure I'm not the only one who bought this and just assumed it would support. I was never able to use it in my projects and I ended up just using it to test some simple PHP logic.

**Comment by mikeerickson** [comment]
@gpressutto5 well, I can say you are not the only one. I do all PHP work using various docker containers.

**Comment by michael-mcmullen** [comment]
> I think they should say on the website that it doesn't support docker. I know they never said it does, but I'm sure I'm not the only one who bought this and just assumed it would support. I was never able to use it in my projects and I ended up just using it to test some simple PHP logic.

I feel like this will be a forgotten project until one is either made open source, or another one comes up. Seems like all the effort is being put towards new projects like expose and the new video course

**Comment by mpociot** [comment]
Forgotten project? We keep putting out updates at least once a month 😅
I know this takes long, but it simply isn’t that easy. I’m working on this, trust me 
And I keep improving Tinkerwell 🙂

**Comment by sschlein** [comment]
@michael-mcmullen Unfortunately, Tinkerwell doesn't pay our bills yet. This is why we need to work on other things in parallel and still do some client work. We are pretty sure to solve this issue within the next months and give Tinkerwell the attention that we want. Running a developer tools company is hard af but we'll make it work :)

**Comment by jessarcher** [comment]
Not sure if it's any help, but I recently did a successful proof of concept for adding Docker support (via `docker-compose`) with auto-detection to zsh-artisan. It's pretty crude and a bit too specific for my taste so I haven't merged it yet, but it's improved my quality of life *drastically*.

https://github.com/jessarcher/zsh-artisan/pull/6

I've got a fair amount of Docker experience, so if there's anything I can do to help here please let me know. I have a copy of Tinkerwell, but sadly I haven't really been able to use it much because of this (I'm not mad - I'm happy to support the amazing work of @mpociot and @sschlein!)

I have no idea how Tinkerwell works under the hood, particularly with SSH, but as others have mentioned if there was some way to customise a prefix for the command I'd imagine that would be a good start. As developers, I think we can naturally handle a bit of tinkering (pun not intended, but I'll take it) with our tools to get things working.

**Comment by mikeerickson** [comment]
@sschlein i have been in this position way to often in my career. Back in late 90s and early 2000s we were selling “widgets” and know how time consuming this can be, and the expectations are from end users who have paid you.

What you are dealing with is a direct correlation to charging money for a commercial product.  This is a different arena compared to open source. I know many many developers who have created commercial offerings in the iOS/Mac AppStore and based on the app, it can be overwhelming. 

Once you put a dollar figure on a product, you assume the responsibility of supporting the product. It is a fine line.

I don’t doubt you are doing a lot to maintain Tinkerwell, but I honestly feel you need to either expand beta releases and get as much variety outside of your machine to help test and track down issues.

A great case and point would be Jetbrains plug-in issue. My guess is that the feature was only really tested on marcel Mac machine and when it worked there, the feature was released.  

Take all this as my 2 cents worth, do with it as you wish but I do have experience so I feel I can add some credible information. I bought Tinkerwell, if for no other reason than to support Marcels efforts.

Good Luck!  Hopefully you will be able to add docker support in the near future and I can use it again as part of my daily workflow

**Comment by mikeerickson** [comment]
@michael-mcmullen well said, I have nothing more to add other than to say I agree.  They are in a precarious position of entering the developer tools market and applying a dollar figure.  Once my money is involved, things are totally different.

**Comment by mpociot** [comment]
I just uploaded a beta version that has Docker support. You can access it via the Beyond Code dashboard at https://beyondco.de/login

Either log in with your existing account or using your license key.

When you open a project that contains a `docker-compose.yaml/yml` file, you will see a new Docker icon in the toolbar. By clicking this, you can configure which docker compose service should be used for running PHP code as well as choosing the working directory within your container (can be auto-detected).

![image](https://user-images.githubusercontent.com/804684/89544046-efebdb00-d801-11ea-902c-9def18df72ee.png)

You can also disable/enable if Docker should be used for the selected project.

Please note: The beta version currently only allows using Docker while in the standard CLI tinker mode. HTTP; Table or Testing mode is not yet modified, as I want to gather some feedback first.

**Comment by herpaderpaldent** [comment]
So hyped to get my hands on this tonight.

**Comment by joelclermont** [comment]
Thank you!!! Just installed it, and it works as expected. Docker toolbar lit up for my project, checked the box, picked the service, and it's working! I'll be using this today and report any issues I run into. Thanks again!

**Comment by sschlein** [comment]
Finally: https://github.com/beyondcode/tinkerwell-community/issues/279

**Comment by gpressutto5** [comment]
@sschlein I don't think this issue should be closed. We now have docker-compose support, not docker support.

**Comment by mpociot** [comment]
@gpressutto5 We want to keep the discussion in #279
